### DISCUSSION TITLE
Allow setting middleware rewrite destination on an existing NextResponse

### DISCUSSION THREAD

**Discussion by zmarty** [vercel/next.js#33477] [Ideas]
### Describe the feature you'd like to request

TL;DR: Please create a helper function which attaches the x-middleware-rewrite header to an existing NextResponse that we previously created with NextResponse.next()

We use [Next.js middleware](https://nextjs.org/docs/middleware) for A/B testing - we send users to one variation of a page or another.

To determine which page (A or B) to serve to a user:
1) We check if a browserId is stored in a cookie called browserId. If not, we generate a GUID with no expiration and set it as the value of the browserId cookie for subsequent visits.
2) Once we know the browserId, we send this semi-permanent browserId to an internal A/B testing endpoint that decides whether to place this user on the A side or B side of an experiment.
3) Rewrite the location to go to either A.tsx or B.tsx page.

Notice that in step 1 we do not know if we are sending the user to page A or B. Also, in this step 1 we sometimes need to **set** a cookie with the browserId. To do that, we need call NextResponse.next() to get an empty response, then we call response.cookie('browserId', ...) on it. 

In steps 2 and 3 we determine where to send the user and at this point we need to call [rewrite](https://github.com/vercel/next.js/blob/d7062dddcc4f7e679cf52214e2a9fc0d87e4f068/packages/next/server/web/spec-extension/response.ts#L89) with the correct path. However, since we started with an empty response that we want to modify, we cannot call NextResponse.rewrite(...) anymore because that would generate a separate new response. What we do instead is a bit of a hack, we manually set the _x-middleware-rewrite_ header on the response we created for step 1.

Since this header name _x-middleware-rewrite_ can presumably change in the future, I propose having an official helper function.

### Describe the solution you'd like

Please create a public helper function which attaches the x-middleware-rewrite header to an existing NextResponse that we previously created with NextResponse.next(). Then have the rewrite function internally call this same helper function.

### Describe alternatives you've considered

We can postpone setting the cookie described in step 1 after we performed steps 2 and 3 but this would be ugly and it would not follow normal logic.

**Comment by balazsorban44** [comment]
We host an [A/B testing example](https://nextjs.org/docs/middleware#examples) which from what you are describing sounds slightly simpler, but also similar. Please have a look:
https://github.com/vercel/examples/blob/main/edge-functions/ab-testing-simple/pages/marketing/_middleware.ts

**Reply by zmarty** [reply]
Your example is essentially the alternative I described, but like I said there this logic is sort of backwards to me.
